doc/mt.1
." DO NOT MODIFY THIS FILE!  It was generated by help2man 1.5.1.3.
.TH MT 1 "June 1999" "mt (Free paxutils) 2.4i" "FSF"
.SH NAME
mt \- manual page for mt (Free paxutils) 2.4i
.SH SYNOPSIS
.B mt
[\fI-V\fR] [\fI-f device\fR] [\fI--file=device\fR] [\fI--help\fR] [\fI--version\fR]\fI operation \fR[\fIcount\fR]
.SH DESCRIPTION
.SH "SEE ALSO"
The full documentation for
.B mt
is maintained as a Texinfo manual.  If the
.B info
and
.B mt
programs are properly installed at your site, the command
.IP
.B info mt
.PP
should give you access to the complete manual.


NEWS
Free paxutils NEWS - User visible changes.	-*- outline -*- (allout)
Copyright © 1995, 1997, 1998, 1999 Free Software Foundation, Inc.

* Version 2.4i - François Pinard, 1999-06.

.* Specification changes
. + An info script should return success for `tar' to continue processing.

.* New features
. + Short `man' pages get installed as well for `cpio', `mt' and `pax'.
. + Option --quick (q) may stop reading the archive early in some cases.
. + Experimental --per-file-compress (y) option may produce safer archives.
. + Using --list (t) with --null produces NUL-terminated, unquoted lines.
. + Options --file-from (T) and --directory (C) may now be used together.
. + Handles buggy Solaris archives having exactly 100 characters in names.

* Version (cpio) 2.4.50 - Tom Tromey, 1998-05.

.* Internationalized.
.* Started pax front end.
.* Test suite now in distribution.

* Version (tar) 1.12b - François Pinard, 1998-02.

.* New features
. + Option --name-prefix for all members of created archive.
. + Experimental --no-attributes option at extraction time.
. + MSDOS port and support, thanks to Eli Zaretskii.
. + A short `man' page gets installed.  `contrib/' holds a better one.

.* Removed features
. + Names --gunzip, --ungzip and --uncompress are obsolete as options.

.* Improvements
. + Reading POSIX `tar' archives, correctly interpret longish file names.
. + Option --owner accepts the USER.GROUP notation.
. + Option --checkpoint writes dots instead of long messages.
. + Various cleanups, speedups, and bug corrections, as usual.

* Version (tar) 1.12 - François Pinard, 1997-04.

.* Sensitive matters
. + Use shell globbing patterns for --label, instead of regular expressions.
. + Do not quote anymore internally over the quoting done by the shell.

.* Output for humans
. + Offer internationalisation capabilities of most recent GNU gettext.
. + Messages available in many more languages, thanks to all translators!
. + Usage of ISO 8601 dates in listings, instead of local American dates.
. + More normalisation and cleanup in error messages.

.* Creation
. + For helping using tar with find, offer a --no-recursion option.
. + Implement --numeric-owner for ignoring symbolic names at create time.
. + New --owner, --group --mode options, still preliminary.
. + Recognise creating an archive on /dev/null, so Amanda works faster.
. + Object to the creation of an empty archive (like in `tar cf FILE').
. + Barely start implementing --posix and POSIXLY_CORRECT.

.* Extraction
. + Make a better job at restoring file and directory attributes.
. + Automatically attempt deleting existing files when in the way.
. + Option --unlink-first (-U) removes most files prior to extraction.
. + Option --recursive-unlink removes non-empty directories when in the way.
. + Option --numeric-owner ignores owner/group names, it uses UID/GID instead.
. + Use global umask when creating missing intermediate directories.
. + When symlinks are not available, extract symbolic links as hard links.
. + Diagnose extraction of contiguous files as regular files.
. + New --backup, --suffix and --version-control options.

.* Various changes
. + Better support of huge archives with --tape-length and --totals.
. + Rename option --read-full-blocks (-B) to --read-full-records (-B).
. + Rename option --block-size (-b) to --blocking-factor (-b).
. + Rename option --record-number (-R) to --block-number (-R).
. + With --block-number (-R), report null blocks and end of file.
. + Implement --record-size for introducing a size in bytes.
. + Delete --block-compress option and rather decide it automatically.
. + Rename option --modification-time to --touch.

.* Many bugs are squashed, while others still run free.

* Version (cpio) 2.4.2 - John Oleynick, 1996-01.

* Version (cpio) 2.3.91 - Tom Tromey, 1995-11.

.* Internationalized.

* Version (cpio) 2.4 - John Oleynick, 1995-06.

.* New texinfo documentation.
.* --sparse option to write sparse files.
.* --only-verify-crc option to verify a CRC format archive.
.* --no-absolute-paths option to ignore absolute paths.
.* --quiet option to supress printing number of blocks copied.
.* Handle disk input errors more gracefully.

* Version (cpio) 2.3 - John Oleynick, 1993-07.

.* in newc and crc format archives, only store 1 copy of multiply linked files.
.* handle multiply linked devices properly.
.* handle multiply linked files with cpio -pl even when the source and
  destination are on different file systems.
.* support HPUX Context Dependent Files.
.* read and write HPUX cpio archives.
.* read System V.4 POSIX tar archives and HPUX POSIX tar archives.
.* use rmdir, instead of unlink, to delete existing directories.

* Version (tar) 1.11.8 - François Pinard, 1995-06.

. + Messages available in French, German, Portuguese and Swedish.
. + The distribution provides a rudimentary Texinfo manual.
. + The device defaults to stdin/stdout, unless overridden by the installer.
. + Option --sparse (-S) should work on more systems.
. + Option --rsh-command may select an alternative remote shell program.

.* Most changes are internal, and should yield better portability.

* Version (tar) 1.11.2 - Michael Bushnell, 1993-03.

.* Changes in backup scripts: cleaned up considerably; notices error
   conditions better over rsh; DUMP_REMIND_SCRIPT is now an option in
   backup-specs; new file dump-remind is an example of a
   DUMP_REMIND_SCRIPT.

.* Superfluous "Reading dirname" was a bug; fixed.

.* Incompatibility problems with a bug on Solaris are fixed.

.* New option --gzip (aliases are --ungzip and -z); calls gzip instead
   of compress.  Also, --use-compress-program lets you specify any
   compress program.  --compress-block is renamed --block-compress and
   now requires one of the three compression options to be specified.

.* Several error messages are cleaned up.

.* Directory owners are now set properly when running as root.

.* Provide DUMP_REMIND_SCRIPT in backup-specs as a possible option
   for --info-script.


src/mt.c
/* mt -- control magnetic tape drive operation
   Copyright (C) 1991, 1992, 1995, 1998, 1999 Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA
   */

/* If -f is not given, the environment variable TAPE is used;
   if that is not set, a default device defined in sys/mtio.h is used.
   The device must be either a character special file or a remote
   tape drive with the form "[user@]system:path".
   The default count is 1.  Some operations ignore it.

   Exit status:
   0	success
   1	invalid operation or device name
   2	operation failed

   Operations (unique abbreviations are accepted):
   eof, weof	Write COUNT EOF marks at current position on tape.
   fsf		Forward space COUNT files.
		Tape is positioned on the first block of the file.
   bsf		Backward space COUNT files.
		Tape is positioned on the first block of the file.
   fsr		Forward space COUNT records.
   bsr		Backward space COUNT records.
   bsfm		Backward space COUNT file marks.
		Tape is positioned on the beginning-of-the-tape side of
		the file mark.
   asf		Absolute space to file number COUNT.
		Equivalent to rewind followed by fsf COUNT.
   eom		Space to the end of the recorded media on the tape
		(for appending files onto tapes).
   rewind	Rewind the tape.
   offline, rewoffl
		Rewind the tape and, if applicable, unload the tape.
   status	Print status information about the tape unit.
   retension	Rewind the tape, then wind it to the end of the reel,
		then rewind it again.
   erase	Erase the tape.

   David MacKenzie <djm@gnu.ai.mit.edu> */

#include "system.h"

#include <getopt.h>
#include "rmt.h"

/* The name this program was run with.  */
char *program_name;

/* If false, don't consider file names that contain a `:' to be
   on remote hosts; all files are local.  Always false for mt;
   since when do local device names contain colons?  */
static bool force_local_option = false;

char *opnames[] =
{
  "eof", "weof", "fsf", "bsf", "fsr", "bsr",
  "rewind", "offline", "rewoffl", "eject", "status",
#ifdef MTBSFM
  "bsfm",
#endif
#ifdef MTEOM
  "eom",
#endif
#ifdef MTRETEN
  "retension",
#endif
#ifdef MTERASE
  "erase",
#endif
  "asf",
#ifdef MTFSFM
  "fsfm",
#endif
#ifdef MTSEEK
  "seek",
#endif
  NULL
};

#define MTASF 600		/* Random unused number.  */
short operations[] =
{
  MTWEOF, MTWEOF, MTFSF, MTBSF, MTFSR, MTBSR,
  MTREW, MTOFFL, MTOFFL, MTOFFL, MTNOP,
#ifdef MTBSFM
  MTBSFM,
#endif
#ifdef MTEOM
  MTEOM,
#endif
#ifdef MTRETEN
  MTRETEN,
#endif
#ifdef MTERASE
  MTERASE,
#endif
  MTASF,
#ifdef MTFSFM
  MTFSFM,
#endif
#ifdef MTSEEK
  MTSEEK,
#endif
  0
};

void
check_type (char *dev, int desc)
{
  struct stat stat_info;

  if (_isrmt (desc))
    return;
  if (fstat (desc, &stat_info) == -1)
    error (1, errno, "%s", dev);
  if ((stat_info.st_mode & S_IFMT) != S_IFCHR)
    error (1, 0, _("%s is not a character special file"), dev);
}

void
perform_operation (char *dev, int desc, short op, int count)
{
  struct mtop control;

  control.mt_op = op;
  control.mt_count = count;
  if (rmtioctl (desc, MTIOCTOP, (char *) &control))
    error (2, errno, "%s", dev);
}

void
print_status (char *dev, int desc)
{
  struct mtget status;

  if (rmtioctl (desc, MTIOCGET, (char *) &status))
    error (2, errno, "%s", dev);

  printf (_("drive type = %d\n"), (int) status.mt_type);
#if defined(hpux) || defined(__hpux__)
  printf (_("drive status (high) = %d\n"), (int) status.mt_dsreg1);
  printf (_("drive status (low) = %d\n"), (int) status.mt_dsreg2);
#else
  printf (_("drive status = %d\n"), (int) status.mt_dsreg);
#endif
  printf (_("sense key error = %d\n"), (int) status.mt_erreg);
  printf (_("residue count = %d\n"), (int) status.mt_resid);
#if !defined(ultrix) && !defined(__ultrix__) && !defined(hpux) && !defined(__hppux) && !defined(__osf__)
  printf (_("file number = %d\n"), (int) status.mt_fileno);
  printf (_("block number = %d\n"), (int) status.mt_blkno);
#endif
}

void
usage (fp, status)
  FILE *fp;
  int status;
{
  fprintf (fp, _("\
Usage: %s [-V] [-f device] [--file=device] [--help] [--version] operation [count]\n"),
	   program_name);
  exit (status);
}

struct option longopts[] =
{
  {"file", 1, NULL, 'f'},
  {"version", 0, NULL, 'V'},
  {"help", 0, NULL, 'H'},
  {NULL, 0, NULL, 0}
};

int
main (int argc, char **argv)
{
  short operation;
  int count;
  char *tapedev;
  int tapedesc;
  int i;

#if DOSWIN
  program_name = get_program_base_name (argv[0]);
#else
  program_name = argv[0];
#endif

  tapedev = NULL;
  count = 1;

  while ((i = getopt_long (argc, argv, "f:t:V:H", longopts, (int *) 0)) != -1)
    {
      switch (i)
	{
	case 'f':
	case 't':
	  tapedev = optarg;
	  break;

	case 'V':
	  printf ("mt (Free %s) %s\n", PACKAGE, VERSION);
	  exit (0);
	  break;

	case 'H':
	default:
	  usage (stdout, 0);
	}
    }

  if (optind == argc)
    usage (stderr, 1);

  i = argmatch (argv[optind], opnames);
  if (i < 0)
    {
      invalid_arg (_("tape operation"), argv[optind], i);
      exit (1);
    }
  operation = operations[i];

  if (++optind < argc)
    count = atoi (argv[optind]);
  if (++optind < argc)
    usage (stderr, 1);

  if (tapedev == NULL)
    {
      tapedev = getenv ("TAPE");
      if (tapedev == NULL)
#ifdef DEFTAPE			/* From sys/mtio.h.  */
        tapedev = DEFTAPE;
#else
	error (1, 0, _("no tape device specified"));
#endif
    }

  if (operation == MTWEOF
#ifdef MTERASE
      || operation == MTERASE
#endif
      )
    tapedesc = rmtopen (tapedev, O_WRONLY, 0, NULL);
  else
    tapedesc = rmtopen (tapedev, O_RDONLY, 0, NULL);
  if (tapedesc == -1)
    error (1, errno, "%s", tapedev);

  check_type (tapedev, tapedesc);

  if (operation == MTASF)
    {
      perform_operation (tapedev, tapedesc, MTREW, 1);
      operation = MTFSF;
    }
  perform_operation (tapedev, tapedesc, operation, count);
  if (operation == MTNOP)
    print_status (tapedev, tapedesc);

  if (rmtclose (tapedesc) == -1)
    error (2, errno, "%s", tapedev);

  exit (0);
}
